From 6055b99ff8001d162cf40dafb2d01ab11e683078 Mon Sep 17 00:00:00 2001
From: Hendrik Brueckner <brueckner@redhat.com>
Date: Mon, 16 Jun 2014 12:55:52 -0400
Subject: [s390] sclp: properly detect line mode console

Message-id: <1402923362-2331-2-git-send-email-brueckner@redhat.com>
Patchwork-id: 84033
O-Subject: [RHEL7.1 PATCH 01/11] [s390] s390/sclp: properly detect line mode console
Bugzilla: 1107730
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>

Description
-----------
commit 22bfda6688c9b91393a07887182ec35acc9d06bb
Author: Christian Borntraeger <borntraeger@de.ibm.com>
Date:   Tue Sep 17 13:28:50 2013 +0200

    s390/sclp: properly detect line mode console

    To detect a line mode console we need a message event type
    for the receive mask and a command event type for the send mask.
    Checking for a MSG event in the sclp send mask was wrong and
    might result in line mode consoles not being detected.

    Signed-off-by: Christian Borntraeger <borntraeger@de.ibm.com>
    Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>

Bugzilla
--------
BZ 1107730
https://bugzilla.redhat.com/show_bug.cgi?id=1107730

Upstream status of the patch
----------------------------
The patch is upstream as of kernel version
http://git.kernel.org/?p=linux/kernel/git/torvalds/linux-2.6.git;a=commitdiff;h=22bfda6688c9b91393a07887182ec35acc9d06bb

Brew
----
https://brewweb.devel.redhat.com/taskinfo?taskID=7572070

Test status
-----------
The patch has been tested and fixes the problem.
The fix has been verified by the IBM test department.

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/s390/char/sclp_cmd.c b/drivers/s390/char/sclp_cmd.c
index 657b034..ae2ca21 100644
--- a/drivers/s390/char/sclp_cmd.c
+++ b/drivers/s390/char/sclp_cmd.c
@@ -145,9 +145,11 @@ bool __init sclp_has_linemode(void)
 
  if (sccb->header.response_code != 0x20)
   return 0;
- if (sccb->sclp_send_mask & (EVTYP_MSG_MASK | EVTYP_PMSGCMD_MASK))
-  return 1;
- return 0;
+ if (!(sccb->sclp_send_mask & (EVTYP_OPCMD_MASK | EVTYP_PMSGCMD_MASK)))
+  return 0;
+ if (!(sccb->sclp_receive_mask & (EVTYP_MSG_MASK | EVTYP_PMSGCMD_MASK)))
+  return 0;
+ return 1;
 }
 
 bool __init sclp_has_vt220(void)
-- 
1.7.1